File: java/IntersectionofTwoLinkedLists.java
/*
 * Filename: IntersectionofTwoLinkedLists.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 18 Jun 2015 02:08:14 PM CST
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
	public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
		if (headA == null || headB == null) {
			return null;
		}

		int lenA = length(headA);
		int lenB = length(headB);
		
		if (lenA > lenB) {
			for (int i = 0; i < lenA - lenB; i++) {
				headA = headA.next;
			}
		} else {
			for (int i = 0; i < lenB - lenA; i++) {
				headB = headB.next;
			}
		}

		ListNode head = new ListNode(0);
		ListNode p = head;

		while (headA != null) {
			if (headA.val == headB.val) {
				p.next = new ListNode(headA.val);
				p = p.next;
			} else {
				head.next = null;
				p = head;
			}
			
			headA = headA.next;
			headB = headB.next;
		}

		return head.next;
	}

	private int length(ListNode head) {
		int len = 0;

		while (head != null) {
			len++;
			head = head.next;
		}

		return len;
	}
}

File: cpp/intersectionOfTwoLinkedLists.cpp
/*
 * Filename: intersectionOfTwoLinkedLists.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 19 Jan 2016 06:38:40 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
	public:
		ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
		{
			ListNode* p = headA;
			ListNode* q = headB;
			ListNode* r;

			while (p && q)
			{
				p = p->next;
				q = q->next;
			}

			if (p)
			{
				r = headA;
				while (p)
				{
					p = p->next;
					r = r->next;
				}

				q = headB;
				while (q && r)
				{
					if (q == r)
					{
						return q;
					}

					q = q->next;
					r = r->next;
				}
			}
			else
			{
				r = headB;
				while (q)
				{
					q = q->next;
					r = r->next;
				}

				p = headA;
				while (p && r)
				{
					if (p == r)
					{
						return p;
					}

					p = p->next;
					r = r->next;
				}
			}

			return NULL;
		}
};
